stdlib/public/core/Duration.swift

```diff
@@ -0,0 +1,206 @@
+    self._attoseconds = _attoseconds
+  }
+  
+  internal init(_seconds: Int64, nanoseconds: Int64) {
+    self = Duration.seconds(_seconds) + Duration.nanoseconds(nanoseconds)
+  }
+  
+  @inlinable
+  public var seconds: Int64 {
+    Int64(_attoseconds / 1_000_000_000_000_000_000)
+  }
+  
+  @inlinable
+  public var nanoseconds: Int64 {
+    let seconds = _attoseconds / 1_000_000_000_000_000_000
+    let nanoseconds = 
+      Int64((_attoseconds - seconds * 1_000_000_000_000_000_000) / 1_000_000_000)
+    return nanoseconds
+  }
+}
+
+extension Duration {
+  @inlinable
+  public static func seconds<T: BinaryInteger>(_ seconds: T) -> Duration {
+    return Duration(_attoseconds: _DoubleWidth<Int64>(seconds) * 
+                                 1_000_000_000_000_000_000)
+  }
+  
+  @inlinable
+  public static func seconds(_ seconds: Double) -> Duration {
+    return Duration(_attoseconds: _DoubleWidth<Int64>(seconds * 
+                                 1_000_000_000_000_000_000))
+  }
+  
+  @inlinable
+  public static func milliseconds<T: BinaryInteger>(
+    _ milliseconds: T
+  ) -> Duration {
+    return Duration(_attoseconds: _DoubleWidth<Int64>(milliseconds) * 
+                                 1_000_000_000_000_000)
+  }
+  
+  @inlinable
+  public static func milliseconds(_ milliseconds: Double) -> Duration {
+    return Duration(_attoseconds: _DoubleWidth<Int64>(milliseconds * 
+                                 1_000_000_000_000_000))
+  }
+  
+  @inlinable
+  public static func microseconds<T: BinaryInteger>(
+    _ microseconds: T
+  ) -> Duration {
+    return Duration(_attoseconds: _DoubleWidth<Int64>(microseconds) * 
+                                 1_000_000_000_000)
+  }
+  
+  @inlinable
+  public static func microseconds(_ microseconds: Double) -> Duration {
+    return Duration(_attoseconds: _DoubleWidth<Int64>(microseconds * 
+                                 1_000_000_000_000))
+  }
+  
+  @inlinable
+  public static func nanoseconds<T: BinaryInteger>(
+    _ nanoseconds: T
+  ) -> Duration {
+    return Duration(_attoseconds: _DoubleWidth<Int64>(nanoseconds) * 
+                                 1_000_000_000)
+  }
+}
+
+extension Duration: Codable {
+  private enum CodingKeys: String, CodingKey {
+    case seconds = "seconds"
+    case nanoseconds = "nanoseconds"
+  }
+  
+  public init(from decoder: Decoder) throws {
+    let container = try decoder.container(keyedBy: CodingKeys.self)
+    let seconds = try container.decode(Int64.self, forKey: .seconds)
+    let nanoseconds = try container.decode(Int64.self, forKey: .nanoseconds)
+    self.init(_seconds: seconds, nanoseconds: nanoseconds)
+  }
+  
+  public func encode(to encoder: Encoder) throws {
+    var container = encoder.container(keyedBy: CodingKeys.self)
+    try container.encode(seconds, forKey: .seconds)
+    try container.encode(nanoseconds, forKey: .nanoseconds)
+  }
+}
+
+extension Duration: Hashable { }
+
+extension Duration: Equatable {
+  @inlinable
```

Comment:
Do we really need this (and `<`, `.zero`, `+` etc) to be inlinable? 

(If it's for performance reasons, we can replicate some of those benefits with `@_effects` attributes, although we need to tread carefully.)

I think it'd be preferable if we could leave these opaque, so that we remain able to replace the stand-in for `Int128` with the actual public type when/if it gets implemented.

Note: in contrast to `Duration`, operations on the two system clocks instants aren't inlinable.

Reply:
This got resolved in 71cc760de21c7ffcac240dc78cb927ccd2097cd4